### ISSUE TITLE
"data type not understood" with numpy 1.12.1

### ISSUE THREAD

**Issue by yarikoptic** [statsmodels/statsmodels#3841]
statsmodels 0.8.0 
Tests are failing in Debian unstable with numpy 1.12.1 e.g
```shell
======================================================================
ERROR: test_mixed_recarray (statsmodels.tsa.tests.test_tsa_tools.TestAddTrend)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/build/statsmodels-0.8.0/debian/python-statsmodels/usr/lib/python2.7/dist-packages/statsmodels/tsa/tests/test_tsa_tools.py", line 528, in test_mixed_recarray
    added = tools.add_trend(ra, trend='ct')
  File "/build/statsmodels-0.8.0/debian/python-statsmodels/usr/lib/python2.7/dist-packages/statsmodels/tsa/tsatools.py", line 121, in add_trend
    x = x.astype(np.dtype(descr))
TypeError: data type not understood
```
and in detail:
```
# nosetests --pdb -s -v /build/statsmodels-0.8.0/debian/python-statsmodels/usr/lib/python2.7/dist-packages/statsmodels/tsa/tests/test_tsa_tools.py                   
/build/statsmodels-0.8.0/debian/python-statsmodels/usr/lib/python2.7/dist-packages/statsmodels/compat/pandas.py:56: FutureWarning: The pandas.core.datetools module is deprecated and will be removed in a future version. Please use the pandas.tseries module instead.
  from pandas.core import datetools
test_array (statsmodels.tsa.tests.test_tsa_tools.TestAddTrend) ... ok
test_dataframe (statsmodels.tsa.tests.test_tsa_tools.TestAddTrend) ... ok
test_dataframe_duplicate (statsmodels.tsa.tests.test_tsa_tools.TestAddTrend) ... ok
test_duplicate_const (statsmodels.tsa.tests.test_tsa_tools.TestAddTrend) ... ok
test_mixed_recarray (statsmodels.tsa.tests.test_tsa_tools.TestAddTrend) ... > /build/statsmodels-0.8.0/debian/python-statsmodels/usr/lib/python2.7/dist-packages/statsmodels/tsa/tsatools.py(121)add_trend()
-> x = x.astype(np.dtype(descr))
(Pdb) p descr
[('c0', '<f8'), ('c1', '|i1'), ('c2', '|S4'), (u'const', '<f8'), (u'trend', '<f8')]
(Pdb) p x
rec.array([( 1. , 1, 'aaaa',  1.,  1.), ( 1.1, 2, 'bbbb',  1.,  2.)], 
          dtype=[(u'c0', '<f8'), (u'c1', 'i1'), (u'c2', 'O'), (u'const', '<f8'), (u'trend', '<f8')])
(Pdb) p x.astype(np.dtype(descr))
*** TypeError: TypeError('data type not understood',)
(Pdb) import numpy
(Pdb) numpy.__version__
'1.12.1'
```
anyone ran into this one?  by foo on numpy structured entries and types is not that strong to see immediately what is going on

**Comment by josef-pkt** [comment]
what's the pandas version in that Debian? 

We had this error a few months ago, but IIUC the main fix was in pandas.
I don't think we have a direct workaround in statsmodels, but there is one open PR for related structured dtypes/recarray handling.
(main comments we get is: drop recarray support)

**Comment by yarikoptic** [comment]
```shell
(Pdb) pandas.__version__
'0.20.3'
```
so the most recent release AFAIK

**Comment by josef-pkt** [comment]
based on the file history, it looks like this PR #3348 included fixes for recarray problems, besides other pandas/numpy compat fixes

@bashtage  @TomAugspurger 

@yarikoptic  The plan is to release statsmodels 0.9 end of august, beginning of september. I am not planning on a 0.8.1 release.
I assume something like the above PR could be used to patch the Debian version. If you have problem getting the next statsmodels into Debian, then we could make a 0.8.1 source/informal/tag release with just compat fixes.

**Comment by yarikoptic** [comment]
unfortunately that one c86c14a830b99a1ee4c2cb9802e43597b956c604 doesn't apply cleanly over 0.8.0 so would need to dig pieces out may be later...
I guess I could just wait for 0.9.0, just that currently testing "Release" would not have statsmodels for a while then... but it is nothing major

**Comment by josef-pkt** [comment]
Can you just silence the unit tests?  Using recarrays is mostly an obsolete feature, given the progress that pandas has made.

**Comment by josef-pkt** [comment]
@yarikoptic   BTW: Is that the only numpy/pandas compatibility failure? 
I thought we fixed more, but maybe those are still in deprecation stage.

**Comment by yarikoptic** [comment]
here is the full list of failures I am observing atm 
```
======================================================================
ERROR: test_mixed_recarray (statsmodels.tsa.tests.test_tsa_tools.TestAddTrend)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/build/statsmodels-0.8.0/debian/python-statsmodels/usr/lib/python2.7/dist-packages/statsmodels/tsa/tests/test_tsa_tools.py", line 528, in test_mixed_recarray
    added = tools.add_trend(ra, trend='ct')
  File "/build/statsmodels-0.8.0/debian/python-statsmodels/usr/lib/python2.7/dist-packages/statsmodels/tsa/tsatools.py", line 121, in add_trend
    x = x.astype(np.dtype(descr))
TypeError: data type not understood

======================================================================
ERROR: test_recarray (statsmodels.tsa.tests.test_tsa_tools.TestAddTrend)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/build/statsmodels-0.8.0/debian/python-statsmodels/usr/lib/python2.7/dist-packages/statsmodels/tsa/tests/test_tsa_tools.py", line 483, in test_recarray
    appended = tools.add_trend(recarray)
  File "/build/statsmodels-0.8.0/debian/python-statsmodels/usr/lib/python2.7/dist-packages/statsmodels/tsa/tsatools.py", line 121, in add_trend
    x = x.astype(np.dtype(descr))
TypeError: data type not understood

======================================================================
ERROR: test_add_lag_drop_insert (statsmodels.tsa.tests.test_tsa_tools.TestLagmat)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/build/statsmodels-0.8.0/debian/python-statsmodels/usr/lib/python2.7/dist-packages/statsmodels/tsa/tests/test_tsa_tools.py", line 243, in test_add_lag_drop_insert
    lag_data = sm.tsa.add_lag(data, 'realgdp', 3, drop=True)
  File "/build/statsmodels-0.8.0/debian/python-statsmodels/usr/lib/python2.7/dist-packages/statsmodels/tsa/tsatools.py", line 206, in add_lag
    usemask=False)
  File "/usr/lib/python2.7/dist-packages/numpy/lib/recfunctions.py", line 638, in append_fields
    dtype=base.dtype.descr + data.dtype.descr)
  File "/usr/lib/python2.7/dist-packages/numpy/ma/extras.py", line 153, in masked_all
    a = masked_array(np.empty(shape, dtype),
TypeError: data type not understood
```
the last one repeats multiple times with different parametrizations
1

**Comment by bashtage** [comment]
The problem is handling of unicode in pandas w.r.t. creating rec arrays.  Should be fixed in pandas.

**Comment by TomAugspurger** [comment]
Here's the issue on pandas: https://github.com/pandas-dev/pandas/issues/16358. It hasn't been resolved yet.

**Comment by josef-pkt** [comment]
So why is statsmodels master not failing anymore with this?
I only paid partially attention to this because the error went away in CI testing.

**Comment by yarikoptic** [comment]
heh, thanks @TomAugspurger !!

**Comment by josef-pkt** [comment]
I was hiding it under the rug so the dirt doesn't show (i.e. keep CI green for more important things)
#3662, original issue #3658

PR #3739 should fix most of it

(Why didn't we drop support for Python 2? ;)

**Comment by josef-pkt** [comment]
fixed in #3739
